### ISSUE TITLE
Grok vector-Jacobian product math

### ISSUE THREAD

**Issue by usevector** [usevector/cudagrad#21]
https://pytorch.org/tutorials/beginner/blitz/autograd_tutorial.html

**Comment by usevector** [comment]
**y** = f(**x**)

J = ( d**y**/ dx1 ... d**y**/dxn  )
**y* is a vector

**Comment by usevector** [comment]
> In [vector calculus](https://en.wikipedia.org/wiki/Vector_calculus), the Jacobian matrix ([/dʒəˈkoʊbiən/](https://en.wikipedia.org/wiki/Help:IPA/English),[[1]](https://en.wikipedia.org/wiki/Jacobian_matrix_and_determinant#cite_note-1)[[2]](https://en.wikipedia.org/wiki/Jacobian_matrix_and_determinant#cite_note-2)[[3]](https://en.wikipedia.org/wiki/Jacobian_matrix_and_determinant#cite_note-3) [/dʒɪ-, jɪ-/](https://en.wikipedia.org/wiki/Help:IPA/English)) of a [vector-valued function](https://en.wikipedia.org/wiki/Vector-valued_function) of several variables is the [matrix](https://en.wikipedia.org/wiki/Matrix_(mathematics)) of all its first-order [partial derivatives](https://en.wikipedia.org/wiki/Partial_derivative).

https://en.wikipedia.org/wiki/Jacobian_matrix_and_determinant

Okay this is not hard, a vector valued function like **y**, is a vection of several variables len(vector) where **x** is our vector). First partial derivatives is saying it's just dy/dx, not d^2y/dx

**Comment by usevector** [comment]
![image](https://github.com/usevector/cudagrad/assets/144763418/d90bc07e-15b1-420e-b651-8f39ea3fa290)

**Comment by usevector** [comment]
![image](https://github.com/usevector/cudagrad/assets/144763418/50c1d08e-1bca-4647-bef6-235505b2fbfe)

**Comment by usevector** [comment]
The wiki has an extra step, it's using nabla to say (d/dx1, ..., d/dxn). Lets get into that.

Okay but the tricky bit is that `f` is a function that maps from real vector space `R^n` to `R^m`, what does that mean, we have `n` real inputs and `m` real outputs. So our x has `n` inputs, remember (**y** = f(**n**)), so here `f` is our map to R^n to R^m, where **y** is the output space. This is not derivatives, this is just a function, any function.

Okay now the fun bit, `f` is not a scalar function, it's a vector valued function producing `m` outputs, which is why wiki has it as **f**. d**f**/dx1 is a vector of `m` derivatives going from 1, ..., to m.

I don't quite get the convention why it needs to be transposed until I get to the first step, but the math is simple enough it's just saying instead of having a column of **f**'s outputs, say f_1, and all it's derivatives of the input x_1 to x_n as columns, transpose it so **f**'s derivatives, say for f_1, are all along the row. That row, being  ∇^Tf_1, maybe but I don't know it would be correct that is the same as  ∇^T**f**.

**Comment by usevector** [comment]
> The Jacobian matrix [represents](https://en.wikipedia.org/wiki/Matrix_(mathematics)#Linear_transformations) the [differential](https://en.wikipedia.org/wiki/Total_derivative) of f at every point where f is differentiable.

This looks harmless but if you click differential, it's not what you think:

> the total derivative of a function f at a point is the best [linear approximation](https://en.wikipedia.org/wiki/Linear_approximation) near this point of the function with respect to its arguments. Unlike [partial derivatives](https://en.wikipedia.org/wiki/Partial_derivative), the total derivative approximates the function with respect to all of its arguments, not just a single one.

![image](https://github.com/usevector/cudagrad/assets/144763418/2a724fcd-14d1-40c4-8f18-1c06cf409c54)

> ## Total Derivative
> 
> You're correct; the concept of the total derivative is important when dealing with functions of multiple variables. Specifically, if you have a scalar function f(x1, x2, ..., xn), the total derivative df is expressed as:
> 
> df = (partial f / partial x1) * dx1 + (partial f / partial x2) * dx2 + ... + (partial f / partial xn) * dxn
> 
> Each term (partial f / partial xi) * dxi represents how much f changes with a small change in xi, keeping all other variables constant. When you sum up these individual contributions, you get the total change in f due to small changes in all the input variables x1, x2, ..., xn.
> 
> This concept is particularly powerful because it allows you to understand how a function changes when you alter multiple variables at once. It's widely used in various fields such as physics, engineering, and machine learning.
> 

> In [mathematics](https://en.wikipedia.org/wiki/Mathematics), the total derivative of a function f at a point is the best [linear approximation](https://en.wikipedia.org/wiki/Linear_approximation) near this point of the function with respect to its arguments. Unlike [partial derivatives](https://en.wikipedia.org/wiki/Partial_derivative), the total derivative approximates the function with respect to all of its arguments, not just a single one.

TODO:

In detail, if h is a [displacement vector](https://en.wikipedia.org/wiki/Displacement_vector) represented by a [column matrix](https://en.wikipedia.org/wiki/Column_matrix), the [matrix product](https://en.wikipedia.org/wiki/Matrix_product) J(x) ⋅ h is another displacement vector, that is the best linear approximation of the change of f in a [neighborhood](https://en.wikipedia.org/wiki/Neighborhood_(mathematics)) of x, if f(x) is [differentiable](https://en.wikipedia.org/wiki/Differentiable_function) at x.[[a]](https://en.wikipedia.org/wiki/Jacobian_matrix_and_determinant#cite_note-5) This means that the function that maps y to f(x) + J(x) ⋅ (y – x) is the best [linear approximation](https://en.wikipedia.org/wiki/Linear_approximation) of f(y) for all points y close to x. The [linear map](https://en.wikipedia.org/wiki/Linear_map) h → J(x) ⋅ h is known as the derivative or the [differential](https://en.wikipedia.org/wiki/Total_derivative) of f at x.

**Comment by usevector** [comment]
Okay so actually you can kinda ignore the transpose thing if we're just denoting it as **J**, well not really, but whatever rows are **f**'s derivative along it's input vector. Then J * h, where h is a column, for every row in **J** we dot with this column input **h** which is displacement vector.

Then what's actually happening, every row of derivatives of f_1, for all x's x_1 to x_n. Is multiplied by a displaement vector where the derivative in df_1/x1, ..., df_m/x_n, specifically it's multiplied by the component that matches the derivative. What does that mean, well it's the total derivative, how much has change in direction x_1, by how much f_1 output changes with regards to x_1 (the df_1/d_x1).

This produces a vector, (m, n) * (n, 1) -> (m, 1) which makes sense because **f** has `m` outputs. 

We sum along the nabla direction, in this case the rows. Which is saying for each of the `m` outputs, how much do the inputs influence f_1 to f_m. At least that's what I think it means. It's like vector derivative at a point I think.

**Comment by usevector** [comment]
Weird that torch docs call it **J^T** but wiki says **J**, exact same output I don't know.

**Comment by usevector** [comment]
Now the thing I need to think about is that wiki example about total derivative was for a vector valued function, what the torch docs are talking about with the scalar loss function.

---

When you're working with a scalar-valued function `f` that maps from R^n to R, you can use a gradient vector to represent the sensitivity of `f` to each input dimension. The total derivative of `f` can be thought of as the dot product between the gradient and a displacement vector `h`.

However, the situation changes when your function `f` maps from R^n to R^m (i.e., the function itself is vector-valued). In this case, you no longer use a simple gradient vector. Instead, you use a Jacobian matrix.

The Jacobian matrix `J` is an m x n matrix where the element J[i][j] is the partial derivative of the i-th component of `f` with respect to the j-th dimension of the input.

To find the total derivative in this setting, you would multiply the Jacobian matrix `J` by your displacement vector `h`. The resulting vector gives you an approximate change in `f` for a given change in input `h`. Specifically, each component of the resulting vector tells you how much each output of `f` will change due to the displacement `h`.

So, if you're interested in how a small change in the input space affects a vector-valued function, you would perform this matrix-vector multiplication with the Jacobian and the displacement vector.

This concept is quite essential in the realms of vector calculus and optimization, particularly when you're dealing with functions that have multiple outputs.

---

Okay I think the core of the vector-Jacobian is that it's taking an input vector because we want to know the vector derivative at a point, and to do that, you need to know the input vector. It's basically the hard version of this:

> To calculate the derivative of f(x) = x^2, you'll first differentiate f(x) with respect to x.
> 
> f'(x) = 2x
> 
> Then, to find the derivative at x = 2, you plug 2 into f'(x).
> 
> f'(2) = 2 x 2 = 4
> 
> So, the derivative of x^2 at x = 2 is 4.
>

**Comment by usevector** [comment]
> the [matrix product](https://en.wikipedia.org/wiki/Matrix_product) J(x) ⋅ h is another displacement vector, that is the best linear approximation of the change of f in a [neighborhood](https://en.wikipedia.org/wiki/Neighborhood_(mathematics)) of x, if f(x) is [differentiable](https://en.wikipedia.org/wiki/Differentiable_function) at x

**Comment by usevector** [comment]
> We know that if v1, ..., vn is a basic of V and T : V -> W is linear, then the values of Tv1, ..., Tvn determine the values of T on aritrary vector in V. As we will soon see, matrices are used as an efficient method of recording the values of the Tvj's in terms of a basis of W.

LADR

**Comment by usevector** [comment]
Okay from the gut.

I think **J** is a map that transforms input vectors, by dotting each derivative of f_1, ..., f_m with an input vector. Which basically says, okay for each of these inputs we will multiply and sum and that how much these inputs make output change. We do number of output times, m. **J** is the vector derivative transformation map in Ryan words.

**Comment by usevector** [comment]
So, to directly answer your question:

- The row vector of derivatives `df1/dx1, ..., df1/dxn` tells you how the first output `f1` changes with respect to the basis vectors of the input space `R^n`.

- The column vector `df1/dx1, ..., dfm/dx1` tells you how each of the outputs `f1, f2, ..., fm` changes with respect to the first basis vector of the input space `R^n`.

Each entry in the Jacobian matrix is essentially an element that bridges the basis of the domain `R^n` and the basis of the codomain `R^m`. It tells us how a tiny movement in a specific input direction influences each component of the output.

**Comment by usevector** [comment]
Oh just one part I forgot, look at the PyTorch docs example. If loss is a scalar function with vector inputs then it's m is 1 and it has a vector jacobian matrix (1, n). Then you can chain rule derivatives to get how loss changes with respect to non direct inputs. And vector-Jacobian is a vector output so you can just keep going.

**Comment by usevector** [comment]
Oh in the pytorch docs J^T is actually the transpose of the wiki J because for example in the first row the denominators are all dx1, so swap n and m in the previous comment
